Clarify an empty object in a VP Token cannot be used to signify an error response (#745)
openid/OpenID4VP: 1.0/openid-4-verifiable-presentations-1_0.md

Reviewer: I think this might actually be specific to DC API? e.g. 'A request for a single document type will reveal the End-User is in possession of that Credential' can I think only happen for DC API.

# Response {#response}

A VP Token is only returned if the corresponding Authorization Request contained a `dcql_query` parameter or a `scope` parameter representing a DCQL Query, as defined in (#vp_token_request).

The Wallet MUST return a VP Token only if it contains a set of
  Presentations that satisfy the requirements of the DCQL query according to
(#dcql_query_lang_processing_rules). If the Wallet does not return any
Presentation, for example, because the End-User did not give consent, the
requested Credentials are not available, or the DCQL query can be satisfied
without returning any Presentation (i.e., all Credential Queries that cannot
be fulfilled are optional), the Wallet MUST NOT return a VP Token; if the
Wallet returns a response, it MUST be an error response as defined in
(#error-response). In particular, an empty VP Token (a JSON object without
any entries) MUST NOT be used to signify an error. Privacy considerations that
apply when returning an error response, in particular with regard to End-User

> consent, are defined in (#error-responses).

Reviewer 2: Should we say something about the error to use here? I'd say `access_denied` is the natural choice.

Reviewer 3: > Should we say something about the error to use here? I'd say `access_denied` is the natural choice.

IMO, there is no need for that.

> A VP Token is only returned if the corresponding Authorization Request contained a `dcql_query` parameter or a `scope` parameter representing a DCQL Query (as defined in #vp_token_request).

Reviewer: Unrelated to this PR, but if I raise a separate PR it would conflict - this renders incorrectly (the #vp_token_request doesn't become a link):

<img width="803" height="158" alt="Image" src="https://github.com/user-attachments/assets/bb06bfdd-4d65-47da-82e4-eaaf21b6c784" />

```suggestion
A VP Token is only returned if the corresponding Authorization Request contained a `dcql_query` parameter or a `scope` parameter representing a DCQL Query, as defined in (#vp_token_request).
```

The Wallet MUST return a VP Token only if it contains a set of

>   Presentations that satisfy the requirements of the DCQL query according to

Reviewer 2: ```suggestion
Presentations that satisfy the requirements of the DCQL query according to
```

A VP Token is only returned if the corresponding Authorization Request contained a `dcql_query` parameter or a `scope` parameter representing a DCQL Query (as defined in #vp_token_request).

The Wallet MUST return a VP Token only if the set of Presentations returned in
the VP Token satisfies the requirements of the DCQL query according to
(#dcql_query_lang_processing_rules). If the Wallet does not return any
Presentation, for example, because the End-User did not give consent, the
requested Credentials are not available, or the DCQL query can be satisfied
without returning any Presentation (i.e., all Credential Queries that cannot
be fulfilled are optional), the Wallet MUST NOT return a VP Token; if the
Wallet returns a response, it MUST be an error response as defined in

> (#error-response). In particular, an empty VP Token (a JSON object without

Reviewer: I think my only remaining is that people might decide to return an error response without getting user consent first. I think there's existing text about this somewhere and we should probably reference that text from here?

Reviewer 2: I guess we should point to https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-1_0-wg-draft.html#section-15.9 ?

Author: Thank you. We should probably move some of the more generic content from 15.9.2 to 15.9 Error Response, and make sure the section applies to both, redirect and DC API flows. We would then need to update the reference to 15.9.2 in Annex A to include the generic and the more specific DC API-related section in the privacy considerations.

A VP Token is only returned if the corresponding Authorization Request contained a `dcql_query` parameter or a `scope` parameter representing a DCQL Query (as defined in #vp_token_request).

The Wallet MUST return a VP Token only if the set of Presentations returned in

> the VP Token satisfies the requirements of the DCQL query according to

Reviewer: Maybe this reads slightly better? I can live with existing wording though.

```suggestion
The Wallet MUST return a VP Token only if it contains a set of
  Presentations that satisfy the requirements of the DCQL query according to
```

Reviewer 4: @jogu I think you've got an extra 'only' in there

Reviewer: I do - fixed, thanks!

When a VP Token is returned, the respective response includes the following parameters:

`vp_token`:
: REQUIRED. A JSON-encoded object subject to the following requirements:

  * Each key MUST be the `id` of a Credential Query in the DCQL query.

  * Each value MUST be an array containing one or more Presentations matching
    the corresponding Credential Query.

>   * When `multiple` is omitted or set to `false`, the array MUST contain exactly

Reviewer: This might make it clearer what's been talked about:

```suggestion
  * When `multiple` was omitted in the DCQL query or set to `false`, the array MUST contain exactly
```

If the Wallet is acting within a trust framework that allows the Wallet to determine whether a Request URI belongs to a certain Client Identifier, the Wallet is RECOMMENDED to validate the Verifier's authenticity and authorization given by the Client Identifier and that the Request URI corresponds to this Verifier. If the link cannot be established in those cases, the Wallet MUST refuse the request.

> ## Error Responses {#privacy-error-responses}

Reviewer: This might break links people have to the 1.0 spec so they no longer jump to the right section.

We could use the existing autogenerated id (`(#error-responses)` ) when linking to it instead?

A Wallet SHOULD NOT return any OpenID4VP protocol errors before obtaining End-User consent, when processing a request containing value matching (to avoid revealing values of claims without consent), or issuer selection (to avoid revealing that the End-User has a Credential from a particular authority). Additionally, the End-User consent protects against undetected, repeated requests to the Wallet.

### `wallet_unavailable` Authorization Error Response {#authorization_error_responsewith_the_wallet_unavailable_error_code}

In the event that another component is invoked instead of the Wallet, the End-User SHOULD be informed and give consent before the invoked component returns the `wallet_unavailable` Authorization Error Response to the Verifier.

### Digital Credential API Error Responses {#privacy-dc-api-error}

> Returning any OpenID4VP protocol error, regardless of content, can reveal additional information about the End-User’s underlying Credentials or Wallet in a way that is unique to the Digital Credentials API since reaching the Wallet can be dependent on a Wallet's ability to satisfy the request. For example, platform implementations could only allow Wallets to be selected that satisfy the request. In this case, OpenID4VP protocol error responses can only be returned by a selected Wallet and would therefore reveal that the End-User is in possession of Credentials that satisfy the request. This is in contrast to other engagement methods, in which the Wallet receives the request before learning if it can be fulfilled. What is revealed by a Wallet in those cases depends on how each individual Wallet processes the request.

Reviewer 5: ```suggestion
Returning any OpenID4VP protocol error, regardless of content, can reveal additional information about the End-User’s underlying Credentials or Wallet in a way that is unique to the Digital Credentials API since reaching the Wallet can be dependent on a Wallet's ability to satisfy the request. For example, platform implementations could allow only Wallets that satisfy the request to be selected. In this case, OpenID4VP protocol error responses can only be returned by a selected Wallet and would therefore reveal that the End-User is in possession of Credentials that satisfy the request. This is in contrast to other engagement methods, in which the Wallet receives the request before learning if it can be fulfilled. What is revealed by a Wallet in those cases depends on how each individual Wallet processes the request.
```
